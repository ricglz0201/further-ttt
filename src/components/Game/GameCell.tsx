import React from 'react';
import {GetLabelProps, getLabel} from 'functions/GameUtils';

interface Props extends GetLabelProps {
  handleClick: (boardNumber: number, cellNumber: number) => void,
}

const GameCell = ({
  board,
  boardNumber,
  cellNumber,
  handleClick
}: Props) => {
  const label = getLabel({board, boardNumber, cellNumber});
  const onClick = React.useCallback(() => {
    handleClick(boardNumber, cellNumber)
  }, [boardNumber, cellNumber, handleClick]);
  const onKeyDown = React.useCallback(({keyCode}) => {
    if(keyCode === 32) {
      onClick();
    }
  }, [onClick])
  return (
    <div
      className={`w-33 ${label}`}
      onClick={onClick}
      onKeyDown={onKeyDown}
      role="button"
      tabIndex={-1}
    >
      {label}
    </div>
  );
};

export default GameCell;
