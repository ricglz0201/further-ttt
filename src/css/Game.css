button {
  background-color: transparent;
}
button.X {
  color: #e7040f;
}
button.O {
  color: #19a974;
}
