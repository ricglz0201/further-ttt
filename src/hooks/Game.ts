import React from 'react';
import {
  changePlayer,
  initialState,
  isOccupied,
  lookForWinner,
} from 'functions/GameUtils'
import {
  Board,
  Game,
  MakeMoveProps,
  Player,
  Winner
} from 'types/Game';

export function useCanClick({ board, currentBoard }: Game) {
  return React.useCallback((boardID, id) => {
    const currentValue = board[boardID][id];
    if (isOccupied(currentValue)) return false;
    return boardID === currentBoard || currentBoard === -1;
  }, [board, currentBoard]);
}

interface UsePVPGameProps {
  game: Game,
  setGame: (game: Game) => void,
}

export function usePVPMove({ game, setGame }: UsePVPGameProps) {
  const { currentPlayer } = game;
  return React.useCallback((props: MakeMoveProps) => {
    setGame({
      ...props,
      currentPlayer: changePlayer(currentPlayer),
    });
  }, [currentPlayer, setGame]);
}

export function usePVPNewGame({ game, setGame }: UsePVPGameProps) {
  return React.useCallback(() => {
    setGame({ ...game, ...initialState() });
  }, [setGame, game]);
}

interface AfterMoveProps {
  board: Board,
  boardNumber: number,
  cellNumber: number,
  moveNumber: number,
  winner: Winner | null,
}

interface UseAfterMoveProps {
  newGame: () => void,
  makeMove: (props: MakeMoveProps) => void,
}

export function useAfterMove({
  newGame,
  makeMove
}: UseAfterMoveProps) {
  return React.useCallback(({
    board,
    boardNumber,
    cellNumber,
    moveNumber,
    winner,
  }: AfterMoveProps) => {
    if (winner) {
      newGame();
    } else if (moveNumber === 81) {
      newGame();
    } else {
      makeMove({ board, currentBoard: cellNumber, moveNumber });
    }
  }, [newGame, makeMove]);
}

export function usePVPGame() {
  const [game, setGame] = React.useState(initialState());
  const { board, currentPlayer, moveNumber } = game;
  const canClick = useCanClick(game);
  const newGame = usePVPNewGame({ game, setGame });
  const makeMove = usePVPMove({ game, setGame });
  const afterMove = useAfterMove({ newGame, makeMove });
  const handleClick = React.useCallback(
    (boardNumber: number, cellNumber: number) => {
      if (canClick(boardNumber, cellNumber)) {
        const newMoveNumber = moveNumber + 1;
        board[boardNumber][cellNumber] = currentPlayer === Player.X ? 1 : -1;
        const winner = lookForWinner(board[boardNumber]);
        afterMove({
          board,
          boardNumber,
          cellNumber,
          moveNumber: newMoveNumber,
          winner,
        });
      }
    }, [afterMove, board, canClick, currentPlayer, moveNumber]);
  return { game, handleClick };
}
