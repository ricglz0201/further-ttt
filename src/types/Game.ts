/**
 * General types for the game folder
 */

// Enums

export enum BoardValue {
  X = 1,
  O = -1,
  Empty = 0,
};

export enum Player {
  X,
  O,
};

// Types

export type BoardRow = [
  BoardValue,
  BoardValue,
  BoardValue,
  BoardValue,
  BoardValue,
  BoardValue,
  BoardValue,
  BoardValue,
  BoardValue,
];

export type Board = [
  BoardRow,
  BoardRow,
  BoardRow,
  BoardRow,
  BoardRow,
  BoardRow,
  BoardRow,
  BoardRow,
  BoardRow,
];

export type BoardLabel = 'X' | 'O' | '';

// Interfaces

export interface Game {
  board: Board,
  currentBoard: number,
  currentPlayer: Player,
  moveNumber: number,
};
