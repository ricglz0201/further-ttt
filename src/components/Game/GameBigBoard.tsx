import React from 'react';
import GameBoardRow from 'components/Game/GameBoardRow';

export const enum BoardValue {
  X = 1,
  O = -1,
  Empty = 0,
}
export type Board = Array<Array<BoardValue>>;

interface Props {
  board: Board,
  currentBoard: number,
  handleClick: (boardNumber: number, cellNumber: number) => void,
}

  const GameBigBoard = (props: Props) => {
  return (
    <div className="flex flex-row jcc">
      <div className="w-75 w-sm-66 w-md-60 w-lg-50">
        {[0, 3, 6].map(initialBoardNumber => (
          <GameBoardRow
            initialBoardNumber={initialBoardNumber}
            {...props}
          />
        ))}
      </div>
    </div>
  );
};

export default GameBigBoard;
