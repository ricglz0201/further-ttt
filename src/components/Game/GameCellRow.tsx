import React from 'react';
import GameCell from 'components/Game/GameCell';
import { Board } from 'types/Game';

interface Props {
  board: Board,
  boardNumber: number,
  currentBoard: number,
  handleClick: (boardNumber: number, cellNumber: number) => void,
  initialCellNumber: number,
}

const GameCellRow = ({ initialCellNumber, ...rest }: Props) => {
  return (
    <>
      {[0, 1, 2].map((index) => (
        <GameCell
          cellNumber={initialCellNumber + index}
          {...rest}
        />
      ))}
    </>
  );
};

export default GameCellRow;
