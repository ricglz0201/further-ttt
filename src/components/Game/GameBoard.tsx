import React from 'react';
import {Board} from 'components/Game/GameBigBoard';
import GameCellRow from 'components/Game/GameCellRow';

interface BoardClassNameProps {
  boardNumber: number,
  currentBoard: number,
}

interface Props extends BoardClassNameProps {
  board: Board,
  handleClick: (boardNumber: number, cellNumber: number) => void,
}

function boardClassName({
  boardNumber,
  currentBoard,
}: BoardClassNameProps) {
  let className = 'w-25';
  const boardIsPlayable =
    boardNumber === currentBoard ||
    currentBoard === -1;
  if(!boardIsPlayable) {
    className += 'bg-grey';
  }
  return className;
}

const GameBoard = ({
  boardNumber,
  currentBoard,
  ...rest
}: Props) => {
  const className = boardClassName({boardNumber, currentBoard});
  return (
    <div className={className}>
        {[0, 3, 6].map(initialCellNumber => (
          <GameCellRow
            boardNumber={boardNumber}
            currentBoard={currentBoard}
            initialCellNumber={initialCellNumber}
            {...rest}
          />
        ))}
    </div>
  );
};

export default GameBoard;
