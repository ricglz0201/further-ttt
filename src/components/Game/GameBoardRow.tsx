import React from 'react';
import GameBoard from 'components/Game/GameBoard';
import { Board } from 'types/Game';

interface Props {
  board: Board,
  currentBoard: number,
  handleClick: (boardNumber: number, cellNumber: number) => void,
  initialBoardNumber: number,
}

const GameBoardRow = ({ initialBoardNumber, ...rest }: Props) => {
  return (
    <>
      {[0, 1, 2].map((index) => (
        <GameBoard
          boardNumber={initialBoardNumber + index}
          {...rest}
        />
      ))}
    </>
  );
};

export default GameBoardRow;
