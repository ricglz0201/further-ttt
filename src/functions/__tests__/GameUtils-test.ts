import {
  BoardClassNameProps,
  GetLabelProps,
  boardClassName,
  getLabel,
  initialState,
  isOccupied,
  changePlayer,
  newBoard,
} from 'functions/GameUtils';
import { Board, BoardValue, BoardLabel, BoardRow, Player } from 'types/Game';

describe('GameUtils', () => {
  describe('boardClassName', () => {
    const argument: BoardClassNameProps = {
      boardNumber: 2,
      currentBoard: -1,
    };
    const backgroundClass = 'bg-grey'
    let result: string;

    afterEach(() => {
      const constantClass = 'w-25'
      expect(result).toEqual(expect.stringContaining(constantClass));
    });

    it('does not contain bg-grey when is the first round', () => {
      result = boardClassName(argument);
      expect(result).toEqual(expect.not.stringContaining(backgroundClass));
    });

    it('does not contain bg-grey when is the current board', () => {
      result = boardClassName({
        ...argument,
        currentBoard: argument.currentBoard,
      });
      expect(result).toEqual(expect.not.stringContaining(backgroundClass));
    });

    it('does contain bg-grey when is not the current board', () => {
      result = boardClassName({
        ...argument,
        currentBoard: 1,
      });
      expect(result).toEqual(expect.stringContaining(backgroundClass));
    });
  })

  describe('getLabel', () => {
    const rowOfzeros: BoardRow = [0, 0, 0, 0, 0, 0, 0, 0, 0];
    const board: Board = [
      [...rowOfzeros],
      [...rowOfzeros],
      [...rowOfzeros],
      [...rowOfzeros],
      [...rowOfzeros],
      [...rowOfzeros],
      [...rowOfzeros],
      [...rowOfzeros],
      [...rowOfzeros],
    ];
    const boardNumber = 1;
    const cellNumber = 2;
    const testCases: Array<[BoardValue, BoardLabel]> = [
      [BoardValue.X, 'X'],
      [BoardValue.O, 'O'],
      [BoardValue.Empty, ''],
    ]

    it.each(testCases)(
      'returns "X" when the boards value is "X"',
      (value: BoardValue, expectedResult: BoardLabel) => {
        board[boardNumber][cellNumber] = value;
        const argument: GetLabelProps = { board, boardNumber, cellNumber };
        expect(getLabel(argument)).toBe(expectedResult);
      }
    );
  })

  describe('isOccupied', () => {
    test.each([BoardValue.X, BoardValue.O])('%s', (value: BoardValue) => {
      expect(isOccupied(value)).toBe(true);
    })

    test('Empty', () => {
      expect(isOccupied(BoardValue.Empty)).toBe(false);
    })
  })

  describe('changePlayer', () => {
    it('changes Player X to O', () => {
      expect(changePlayer(Player.X)).toBe(Player.O);
    });

    it('changes Player O to X', () => {
      expect(changePlayer(Player.O)).toBe(Player.X);
    });
  });

  test('newBoard', () => {
    expect(newBoard()).toMatchSnapshot();
  });

  test('initialState', () => {
    expect(initialState()).toMatchSnapshot();
  });
});
