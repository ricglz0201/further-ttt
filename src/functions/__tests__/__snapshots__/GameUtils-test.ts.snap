// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GameUtils initialState 1`] = `
Object {
  "board": Array [
    Array [
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
    ],
    Array [
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
    ],
    Array [
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
    ],
    Array [
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
    ],
    Array [
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
    ],
    Array [
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
    ],
    Array [
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
    ],
    Array [
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
    ],
    Array [
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
    ],
  ],
  "currentBoard": -1,
  "currentPlayer": 0,
  "moveNumber": 0,
}
`;

exports[`GameUtils newBoard 1`] = `
Array [
  Array [
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
  ],
  Array [
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
  ],
  Array [
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
  ],
  Array [
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
  ],
  Array [
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
  ],
  Array [
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
  ],
  Array [
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
  ],
  Array [
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
  ],
  Array [
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
  ],
]
`;
